### ISSUE TITLE
ERR! leader not found, after connecting to the leader timeout

### ISSUE THREAD

**Issue by jasonbigl** [rqlite/rqlite#1818]
Hi,

today one of non-voter in my cluster stop getting new data from leader, when I'm trying to execute sql on it, it return the error
`ERR! server responded with 503 Service Unavailable: leader not found`

I inspect the non-voter's log and found the following logs:
```
[store] 2024/06/24 19:00:16 incremental snapshot created in 11.503231ms on node ID 10.1.43.105
[raft] 2024/06/24 19:00:16 [INFO]  starting snapshot up to: index=589302
[store] 2024/06/24 19:00:16 persisted snapshot 2-589302-1719226816938 in 2.042189ms
[raft] 2024/06/24 19:00:16 [INFO]  no logs to truncate
[raft] 2024/06/24 19:00:16 [INFO]  snapshot complete up to: index=589302
2024-06-24T19:15:15.156+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:6130: read: connection reset by peer"
[mux] 2024/06/24 20:06:42 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:51178: i/o timeout
[mux] 2024/06/24 20:06:59 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:23798: i/o timeout
2024-06-24T20:08:29.797+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:33432: read: connection timed out"
[mux] 2024/06/24 20:08:48 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:16480: i/o timeout
2024-06-24T20:09:28.933+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:14842: read: connection timed out"
```

the `10.0.45.100` is the leader, is this a expected behavior that the non-voter stop getting new data once it's connection to leader is temporarily down?

The leader is in United states and the non-voter is in Asia, they are connected via AWS internal VPC. the entire cluster have no problem for over 2 months, but today I get this error, may I ask how can I prevent this from happening in the future?

I'm using `version v8.16.6`,  the reason I use this version is I have a unreproduceable issue https://github.com/rqlite/rqlite/issues/1738, and I suspect it's related to https://github.com/rqlite/rqlite/issues/1711, so I roll back and also use v8.16.6, after that I don't have random  "malformed" error anymore.

Thank you!

**Comment by otoolep** [comment]
Are you doing reads or writes through the non voting mode? If writes, then you need a connection to the leader. There is no way around this. 
If reads you need to select NONE or AUTO as your read consistency mode, so that the non voter does not attempt to contact the Leader. 
The system seems to be working exactly as expected. 

https://rqlite.io/docs/api/read-consistency/ 

Sent from Yahoo Mail for iPhone


On Tuesday, June 25, 2024, 2:50 AM, jasonbigl ***@***.***> wrote:




Hi,

today one of non-voter in my cluster stop getting new data from leader, when I'm trying to execute sql on it, it return the error
ERR! server responded with 503 Service Unavailable: leader not found

I inspect the non-voter's log and found the following logs:
[store] 2024/06/24 19:00:16 incremental snapshot created in 11.503231ms on node ID 10.1.43.105
[raft] 2024/06/24 19:00:16 [INFO]  starting snapshot up to: index=589302
[store] 2024/06/24 19:00:16 persisted snapshot 2-589302-1719226816938 in 2.042189ms
[raft] 2024/06/24 19:00:16 [INFO]  no logs to truncate
[raft] 2024/06/24 19:00:16 [INFO]  snapshot complete up to: index=589302
2024-06-24T19:15:15.156+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:6130: read: connection reset by peer"
[mux] 2024/06/24 20:06:42 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:51178: i/o timeout
[mux] 2024/06/24 20:06:59 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:23798: i/o timeout
2024-06-24T20:08:29.797+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:33432: read: connection timed out"
[mux] 2024/06/24 20:08:48 cannot read header byte: read tcp 10.1.43.105:29088->10.0.45.100:16480: i/o timeout
2024-06-24T20:09:28.933+0800 [ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:14842: read: connection timed out"

the 10.0.45.100 is the leader, is this a expected behavior that the non-voter stop getting new data once it's connection to leader is temporarily down?

The leader is in United states and the non-voter is in Asia, they are connected via AWS internal VPC. the entire cluster have no problem for over 2 months, but today I get this error, may I ask how can I prevent this from happening in the future?

I'm using version v8.16.6, the reason I use this version is I have a unreproduceable issue #1738, and I suspect it's related to #1711, so I roll back and also use v8.16.6, after that I don't have random "malformed" error anymore.

Thank you!

**Comment by jasonbigl** [comment]
Hi, thank you for look into this, I don't perform any write to non-voters, and I only use pdo driver to read from sqlite file directly, actually I only execute `.tables` command via rqlite shell and it give `leader not found` error, and I found some data are not synced to that non-voter.

I have terminated that instance after I found this problem, and bring up a new non-voter to join the cluster, the data is consistently synced without any problem so far. I want to confirm one thing:

**will non-voter retry to connect to leader if the non-voter failed to reach leader temporarily?**

And now I still can't perform any query on non-voter, have no idea what happen, can you please give some ideas?

```
Welcome to the rqlite CLI.
Enter ".help" for usage hints.
Connected to http://10.2.34.135:29087 running version v8.16.6
10.2.34.135:29087> .tables
[client] 2024/06/29 22:28:40 host '10.2.34.135:29087' is unavailable, retrying with the next available host
[client] 2024/06/29 22:28:40 host '' is unavailable, retrying with the next available host
[client] 2024/06/29 22:28:40 none of the available hosts are responsive
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x28417c]

goroutine 1 [running]:
main.main.func1(0x4000116420)
        /tmp/tmp.PS3SrAAll4/src/github.com/rqlite/rqlite/cmd/rqlite/main.go:233 +0x100c
github.com/mkideal/cli.(*Command).RunWith(0x40000006c0, {0x4000020090, 0x3, 0x3}, {0x0?, 0x0?}, {0x0, 0x0}, {0x0, 0x0, ...})
        /tmp/tmp.PS3SrAAll4/pkg/mod/github.com/mkideal/cli@v0.2.7/command.go:219 +0x2e0
github.com/mkideal/cli.(*Command).Run(...)
        /tmp/tmp.PS3SrAAll4/pkg/mod/github.com/mkideal/cli@v0.2.7/command.go:176
github.com/mkideal/cli.RunWithArgs({0x2bbc80?, 0x4000116370}, {0x4000020080, 0x4, 0x4}, 0x348aa8, {0x0?, 0x49470?, 0x541098?})
        /tmp/tmp.PS3SrAAll4/pkg/mod/github.com/mkideal/cli@v0.2.7/cli.go:30 +0x1e0
github.com/mkideal/cli.Run(...)
        /tmp/tmp.PS3SrAAll4/pkg/mod/github.com/mkideal/cli@v0.2.7/cli.go:15
main.main()
        /tmp/tmp.PS3SrAAll4/src/github.com/rqlite/rqlite/cmd/rqlite/main.go:76 +0x70
```

the rqlite is bring up via command
```
rqlited -auth /rqlite/auth.json -raft-non-voter=true -node-id 10.2.34.135 -join rqlite-main.dbcluster.internal:29088 -join-as super -raft-cluster-remove-shutdown=true -http-addr 0.0.0.0:29087 -http-adv-addr 10.2.34.135:29087 -raft-addr 10.2.34.135:29088 -on-disk-path /data/sqlite.db /raft
```

I connect rqlite by
```
rqlite --host=10.2.34.135 --port=29087 --user=super:PASSWORD
```

the ip:host is reachable based on telnet result
```
telnet 10.2.34.135 29087
Trying 10.2.34.135...
Connected to 10.2.34.135.
Escape character is '^]'.
```

**Comment by otoolep** [comment]
Looks like a bug in the rqlite shell, where it's not handling the state you got your cluster into. I'll take a look.

**Comment by jasonbigl** [comment]
> As for your original report on this issue, I see nothing wrong. The node is complaining it can't contact the leader. Looks like something wrong with your network.

yeah the node is complaining about network, the connection is from Singapore to US via aws vpc peer connection. I want to ask does the node have some kind of retry mechanism? retry connecting to leader after failure.

Because I have talked to aws support and they said this network connection issue is temporarily and suggest I retry the connection.

**Comment by otoolep** [comment]
Assuming we're just talking about reads here (if you show me the actual queries you are sending to the read-only node it would be clearer) you need to have your client check for errors, and implement your own "try the query again" logic.

**Comment by jasonbigl** [comment]
> Assuming we're just talking about reads here (if you show me the actual queries you are sending to the read-only node it would be clearer) you need to have your client check for errors, and implement your own "try the query again" logic.

thank you for the quick reply! maybe I haven't described it clearly, let me try again :)

The problem is the nonvoter stop replicating data after it failed to reach the leader, as long as I found error log like this:
`[ERROR] raft-net: failed to decode incoming command: error="read tcp 10.1.43.105:29088->10.0.45.100:14842: read: connection timed out"`

then that nonvoter won't replicate data from leader any more, so the data in that node is outdated. I want to ask that **will the rqlite retry to contact the leader and then start replicate data again after network failure**? not the query retry, actually I'm directly query the sqlite file, the query is never a problem, just the data in the nonvoter is behind the leader.

**Comment by otoolep** [comment]
If the network heals the read-only node should receive all the updates it missed, and be back to normal within a few seconds. This is a core part of how rqlite works.

If that is not what you're seeing I would need clear evidence that a) the network between the read-only node the Leader is operating perfectly, and b) you're seeing those messages being printed in the log at the same time complaining about network issues, or that c) you can clearly show that the read-only nodes is behind.

**Comment by jasonbigl** [comment]
good to know! thank you very much, I'll keep watching and post all information if I see this again

**Comment by otoolep** [comment]
On other words the simplest answer for what you're seeing is that the network is the problem, and is continually having problems.

Have you tried restarting the read-only node when it gets into this state? If it is a bug in rqlite (always possible, of course) then a restart would probably get past this issue, and suddenly the node would be "up to date".

Anyway, how do you know the node hasn't received all the data from the leader?

I'm going to need more precise details, including how you it's not receiving all data from the leader.

**Comment by otoolep** [comment]
When you say the read-only node is not getting all the data, please HTTP GET `/status` and `/debug/vars` and add it here.

**Comment by jasonbigl** [comment]
got it

1. `how do you know the node hasn't received all the data from the leader?`
I directly query the sqlite file in nonvoter, can't find some records, but if I query the same sql in the leader, the records exist. and those records are created hour ago(based on the created_at column info that I have in the table)

2. `When you say the read-only node is not getting all the data, please HTTP GET /status and /debug/vars and add it here.`
I was trying to collect such info, but when I try to perform `.tables` via rqlite shell, it return `leader not found`, haven't tried to curl the `/status` endpoint, will do. but I assume the curl will return the same since the rqlite shell is the wrapper of curl if I understand right

3. `is that the network is the problem, and is continually having problems.`
When i saw this issue, I tried to perform the `telnet` command to leader in the non voter, the telnet result shows the nonvoter can succesfully connect to the leader raft port, but I didn't try to connect the port described from logs, based on the logs, it seems the nonvoter are trying to connect to leader on various port, not just raft port.

4. `Have you tried restarting the read-only node when it gets into this state`
I can't just restart it, it's in our production cluster, so I perform the blue-green deployment to replace the node. we were happy with rqlite for months, but recently we saw this issue and are trying to figure it out, our rqlite database is pretty small, like ~200 MB

**Comment by otoolep** [comment]
The shell is not a wrapper around `curl`, it has its own logic (which is clearly buggy in some cases). Hit the API on the rqlite node directly using `curl`.

>it seems the nonvoter are trying to connect to leader on various port, not just raft port.

A non-voting node only contacts the Raft port. There is no inter-node network traffic sent to any other port.

>I can't just restart it, it's in our production cluster, 

Well, as an aside, that's a fairly brittle production system in that case. Nodes can crash at anytime, exit, and so on. You already say it's having trouble contacting the Leader. If so, it's close to being offline.

**Comment by otoolep** [comment]
>I'm using version v8.16.6, the reason I use this version is I have a unreproduceable issue [#1738](https://github.com/rqlite/rqlite/issues/1738), and I suspect it's related to [#1711](https://github.com/rqlite/rqlite/issues/1711), so I roll back and also use v8.16.6, after that I don't have random "malformed" error anymore.

I would be very interested in learning more about this. If you can show me that an older version of rqlite is not getting "malformed" but a newer version is, then I need to fix that (or at least explain why).

The issue you referenced doesn't make sense. [#1711](https://github.com/rqlite/rqlite/issues/1711) doesn't really make sense as a possible cause.

**Comment by jasonbigl** [comment]
yes I agree, I have added health check in our system so if the node is offline, that instance will be removed from system

as for the random "malformed" error, I just can't reproduce it, not sure if it's the sqlite issue or it's on the rqlite side, still trying to find a stable way to reproduce it, but are happy with v8.16.6

**Comment by otoolep** [comment]
Going to close this now, due to lack of activity.
